src/dhcpmon/src/dhcp_device.c

```diff
@@ -254,34 +386,90 @@ static bool dhcp_device_is_dhcp_inactive(uint64_t counters[][DHCP_DIR_COUNT][DHC
  *
  * @return true if DHCP message 'type' is transmitted,false otherwise
  */
-static bool dhcp_device_is_dhcp_msg_unhealthy(dhcp_message_type_t type,
-                                              uint64_t counters[][DHCP_DIR_COUNT][DHCP_MESSAGE_TYPE_COUNT])
+static bool dhcp_device_is_dhcpv4_msg_unhealthy(dhcpv4_message_type_t type,
+                                                uint64_t v4counters[][DHCP_DIR_COUNT][DHCPv4_MESSAGE_TYPE_COUNT])
 {
     // check if DHCP message 'type' is being relayed
-    return ((counters[DHCP_COUNTERS_CURRENT][DHCP_RX][type] >  counters[DHCP_COUNTERS_SNAPSHOT][DHCP_RX][type]) &&
-            (counters[DHCP_COUNTERS_CURRENT][DHCP_TX][type] <= counters[DHCP_COUNTERS_SNAPSHOT][DHCP_TX][type])    );
+    return ((v4counters[DHCP_COUNTERS_CURRENT][DHCP_RX][type] >  v4counters[DHCP_COUNTERS_SNAPSHOT][DHCP_RX][type]) &&
+            (v4counters[DHCP_COUNTERS_CURRENT][DHCP_TX][type] <= v4counters[DHCP_COUNTERS_SNAPSHOT][DHCP_TX][type])    );
+}
+
+/**
+ * @code dhcp_device_is_dhcpv6_msg_unhealthy(type, counters);
+ *
+ * @brief Check if DHCP relay is functioning properly for message of type 'type'.
+ *        For every rx of message 'type', there should be increment of the same message type.
+ *
+ * @param type      DHCP message type
+ * @param counters  current/snapshot counter
+ *
+ * @return true if DHCP message 'type' is transmitted,false otherwise
+ */
+static bool dhcp_device_is_dhcpv6_msg_unhealthy(dhcpv6_message_type_t type,
+                                                uint64_t v6counters[][DHCP_DIR_COUNT][DHCPv6_MESSAGE_TYPE_COUNT])
+{
+    // check if DHCP message 'type' is being relayed
+    switch (type) {
+        case DHCPv6_MESSAGE_TYPE_SOLICIT:
+        case DHCPv6_MESSAGE_TYPE_REQUEST:
+            return ((v6counters[DHCP_COUNTERS_CURRENT][DHCP_RX][type] > v6counters[DHCP_COUNTERS_SNAPSHOT][DHCP_RX][type]) &&
+                    (v6counters[DHCP_COUNTERS_CURRENT][DHCP_TX][DHCPv6_MESSAGE_TYPE_RELAY_FORWARD] <= v6counters[DHCP_COUNTERS_SNAPSHOT][DHCP_TX][DHCPv6_MESSAGE_TYPE_RELAY_FORWARD]));
+
+        case DHCPv6_MESSAGE_TYPE_RELAY_REPLY:
```

Comment:
I believe the whole `switch` case could be eliminate.

Reply:
can you explain why we can eliminate the switch case here?
The check is dependent on the packet type, for solicit/request packets we need to check relay-forward packet.
For relay-reply packet we need to check advertise/reply packet.
Are you agree or have a better idea?

Reply:
Addressed on a different PR.